### PR TITLE
feat(pipeline executions/orca) : Added ability to add roles to manual judgment stage.

### PR DESCRIPTION
**Pull request by sanopsmx**
feat(pipeline executions/orca) : Added ability to add roles to manual judgment stage.

This is part of: spinnaker/spinnaker#4792.

Enhanced ManualJudgmentStage.groovy to

Get the application roles, stage roles and the user roles.
Check each of the user roles whether they are contained in the stage and application roles.
Once the user role is contained in the manual judgment stage.
We check for whether the stage role has 'READ', 'WRITE', 'EXECUTE', 'CREATE' role in the application permissions,
if the role has application permission as 'READ', then the user will not be allowed to proceed further to subsequent(downstream) stages.
if the role has application permission as 'WRITE, EXECUTE,CREATE', then the user will be allowed to proceed further to subsequent stages.
If yes/no, set a isAuthorized flag to true/false by checking the userRoles, stageRoles and application Roles.

If yes/no, continue with the next stages/continues running the same stage.

Enhanced ManualJudgmentStageSpec.groovy to

Modified the testcases as per the requirement.

Added ManualJudgmentAuthzGroupsUtil.groovy

Added a utlity method to check authorized groups of the manual judgment stage.

Added ManualJudgmentAuthzGroupsUtilSpec.groovy

Added test cases check authorized groups method of ManualJudgmentAuthzGroupsUtil.groovy.

### CODE REVIEW

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy [hunk 1]
```diff
@@ -95,12 +112,32 @@ class ManualJudgmentStage implements StageDefinitionBuilder, AuthenticatedStage
...
+        notificationState = "manualJudgment"
+        executionStatus = ExecutionStatus.RUNNING
+        stage.context.put("judgmentStatus", "")
+      }
       Map outputs = processNotifications(stage, stageData, notificationState)
 
       return TaskResult.builder(executionStatus).context(outputs).build()
     }
 
+    boolean checkManualJudgmentAuthorizedGroups(def stageRoles, def permissions, def username) {
```
Comment (ajordens) [inline orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:125]: Please type these parameters (even if it's groovy)

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy [hunk 2]
```diff
@@ -72,21 +75,35 @@ class ManualJudgmentStage implements StageDefinitionBuilder, AuthenticatedStage
...
+      this.fiatPermissionEvaluator = fpe.orElse(null)
+    }
 
     @Override
     TaskResult execute(StageExecution stage) {
       StageData stageData = stage.mapTo(StageData)
+      def stageAuthorized = stage.context.get('isAuthorized')
+      def stageRoles = stage.context.get('stageRoles')
+      def permissions = stage.context.get('permissions')
+      def username = stage.lastModified? stage.lastModified.user: stage.context.get('username')
```
Comment (ajordens) [inline orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:93]: I think you should be able to fetch the user from `AuthenticatedRequest.getSpinnakerUser()`.

Defaulting to `stage.context.get('username')` seems suspicious in this case.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy [hunk 3]
```diff
@@ -204,6 +205,52 @@ class OperationsController {
     }
   }
 
+  private void addStageAuthorizedRoles(def request, Map pipeline) {
```
Comment (ajordens) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy:208]: I don't understand what this is doing, or why it's doing it when we're starting a pipeline.

Is the idea not that a manual judgment stage can be configured with a set of roles needed to judge it?

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy [hunk 4]
```diff
@@ -72,21 +75,35 @@ class ManualJudgmentStage implements StageDefinitionBuilder, AuthenticatedStage
...
+
+    WaitForManualJudgmentTask(Optional<EchoService> echoService, Optional<FiatPermissionEvaluator> fpe) {
+      this.echoService = echoService.orElse(null)
+      this.fiatPermissionEvaluator = fpe.orElse(null)
+    }
 
     @Override
     TaskResult execute(StageExecution stage) {
       StageData stageData = stage.mapTo(StageData)
+      def stageAuthorized = stage.context.get('isAuthorized')
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:90]: done.

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 1 above)
Comment (sanopsmx) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:125]: done

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 2 above)
Comment (sanopsmx) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:93]: I should have deleted it. Kept it for executing the test cases. Modified the test case.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy (hunk 3 above)
Comment (sanopsmx) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy:208]: This method will check for manual judgment stages and will find out if the user is authorized to continue. Will set the isAuthorized flag to true , if authorized,

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 2 above)
Comment (ajordens) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy]: I still don't believe accessing `stage.lastModified` directly is correct, what does AuthenticatedRequest.getSpinnakerUser() return? 

I _believe_ it should return the correct user.

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy [hunk 5]
```diff
@@ -72,21 +76,34 @@ class ManualJudgmentStage implements StageDefinitionBuilder, AuthenticatedStage
     final long backoffPeriod = 15000
     final long timeout = TimeUnit.DAYS.toMillis(3)
 
-    @Autowired(required = false)
-    EchoService echoService
+    private final EchoService echoService
+
+    private final FiatPermissionEvaluator fiatPermissionEvaluator
+
+    WaitForManualJudgmentTask(Optional<EchoService> echoService, Optional<FiatPermissionEvaluator> fpe) {
```
Comment (ajordens) [inline orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:83]: No `@Autowired` annotation needed on this constructor?

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy (hunk 3 above)
Comment (ajordens) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy:208]: It's still unclear why it needs to happen in the `OperationsController`. I'm confused as this code seems to be manipulating every stage in the pipeline?

Overall this implementation is hard to follow.

My expectation was that the manual judgment stage would have a set of required roles in its context, and would check against those roles when determining whether or not to continue. That shouldn't require changes anywhere other than the manual judgement stage/task.

As an aside, there is a small loophole here in that any stage context can be PATCH'd. See https://github.com/spinnaker/orca/blob/master/orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy#L475.

File: orca-echo/src/test/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStageSpec.groovy [hunk 6]
```diff
@@ -27,31 +30,67 @@ import static com.netflix.spinnaker.orca.echo.pipeline.ManualJudgmentStage.Notif
...
+    then:
+    1 * fpe.getPermission('abc@somedomain.io') >> {
+      new UserPermission().addResources([new Role('foo'), new Role('baz')]).view
+    }
+    result.status == expectedStatus
+
+    where:
+    context                      || expectedStatus
+    [judgmentStatus: "continue", isAuthorized: false, stageRoles: ['foo'],
+     permissions: [WRITE: ["foo"],READ: ["foo","baz"], EXECUTE: ["foo"]]] || ExecutionStatus.SUCCEEDED
```
Comment (ajordens) [inline orca-echo/src/test/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStageSpec.groovy:78]: What is the difference between `permissions` and `stageRoles`? I thought the intention was you needed to have a particular role (or roles?). 

The reason for WRITE/READ/EXECUTE isn't clear.

Comment (sanopsmx) [reply orca-echo/src/test/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStageSpec.groovy:78]: stageRoles --> Roles configured at  the manual judgment stage.
permissions --> roles configured during the application.creation.

Please refer the design document table for more clear picture.
https://github.com/spinnaker/governance/pull/80

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 5 above)
Comment (sanopsmx) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:83]: There is no autowired annotation on this constructor. Please check the file again.

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 2 above)
Comment (sanopsmx) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy]: stage.lastModified returns the user, who clicked on the 'continue' or 'stop' button of the manual judgment stage.

this should work. it returns the correct user. 

If you still insist, i can change the code.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy (hunk 3 above)
Comment (sanopsmx) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy:208]: It doesn't manipuate every stage. It only manipulates manual judgment stages.

Please refer the design document.
https://github.com/spinnaker/governance/pull/80

Comment (ajordens) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy:208]:
```
def stageList = pipeline.stages
            def stageRoles = []
            stageList.each { item ->
              stageRoles = item.selectedStageRoles
              item.isAuthorized = ManualJudgmentAuthzGroupsUtil.checkAuthorizedGroups(userRoles, stageRoles, permissions)
              item.stageRoles = stageRoles
              item.permissions = permissions
            }
```

What am I missing here, that's iterating across `pipeline.stages`.

Comment (sanopsmx) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy:208]: item.selectedStageRoles returns null for all the stages except manual judgment(if configured).
This decision to apply for all stages was taken in our design meet, so that in future, if we want to extend it to all stages, this will work.

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 5 above)
Comment (ajordens) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:83]: I realize there isn't, but this class was changed from field injection to constructor injection and for consistency we usually add an `@Autowired` annotation to the constructor.

File: orca-echo/src/test/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStageSpec.groovy (hunk 6 above)
Comment (ajordens) [reply orca-echo/src/test/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStageSpec.groovy:78]: > Display all the groups the user has selected while creating application(permissions) in the authorized groups dropdown.

Sure. But that set of application permissions should just be looked up from the UI, it doesn't need to be stored in the stage context. 

I'd like this PR to really be centered around the managed delivery stage and required user roles needed to act on it.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy (hunk 3 above)
Comment (ajordens) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy:208]: If we wanted to apply similar logic to all stages, we would do so with a different implementation (likely a preprocessor that could run before task execution).

`selectedStageRoles` may be null, but you're also manipulating the context for every stage with `stageRoles` and `permissions`.

It also introduces another remote call in the flow to run a pipeline (to `front50`) which is even with this feature as currently implemented would be unnecessary to anyone who isn't using roles in manual judgment stages. As a general rule, we try to remove unnecessary rpc calls on critical paths.

I'd like this changed, and really this bit removed.

**Review by ajordens — CHANGES_REQUESTED** [review]

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 2 above)
Comment (ajordens) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy]: Yes, `lastModified` is an implementation detail. My preference would be to avoid referencing it directly.

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 5 above)
Comment (ajordens) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:83]: Most of the project is going to have `@Autowired` on `@Component` constructors. I believe it works in this case because the class has only one constructor, but the norm as far as I'm concerned is to annotate the constructors even if there is only one.

This is a minor point on this PR.

File: orca-echo/src/test/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStageSpec.groovy (hunk 6 above)
Comment (sanopsmx) [reply orca-echo/src/test/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStageSpec.groovy:78]: I am unable to follow you here. Did not understand what you meant? Can you be more specific.

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 5 above)
Comment (sanopsmx) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:83]: Should i use @Autowired or not? Since, this is a minor point. I am dropping this feedback.

File: orca-echo/src/test/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStageSpec.groovy (hunk 6 above)
Comment (ajordens) [reply orca-echo/src/test/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStageSpec.groovy:78]: Sorry, for some odd reason I typed "managed delivery" when I meant "manual judgment". I blame the late hour, sorry :(

I think what would be clearer is if I did a quick prototype/pr of what is in my mind, and we can compare. I'll do that this morning.

**Comment by ajordens** [comment]
Spent a few minutes this morning and put together https://github.com/spinnaker/orca/pull/3991.

How different are the intentions here than that? 

The `PATCH` call already verifies that the user has `EXECUTE` against the application, so I don't think any further validation is needed.

I did mention earlier that this only works in the happy path but that `PATCH` operation does allow for the set of required roles to be overwritten. I believe a more legitimate fix for that would be to split immutable inputs out of the stage context. No timeline on when that will happen.

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 5 above)
Comment (robzienert) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:83]: I asked you to remove the `@Autowired` properties in favor of a [single constructor](https://github.com/spinnaker/orca/pull/3947/files#r513595076). My example did not include an `@Autowired` annotation, but that wasn't really intended to be copy/pasted, either. While `@Autowired` on a single constructor is optional, per @ajordens, precedent in the rest of the codebase does have us adding `@Autowired` on single constructors anyway, probably for explicitness.

This is a good example for why keeping a single PR for a change is valuable: I had to click through 4 different PRs to find the original thread of feedback about this. Had it been a single PR, context for other reviewers would've been easily discoverable.

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 2 above)
Comment (sanopsmx) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy]: done...fetching user from AuthenticatedRequest.getSpinnakerUser().

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy (hunk 3 above)
Comment (sanopsmx) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy:208]: removed this piece of code completely now.

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 5 above)
Comment (sanopsmx) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:83]: done..added @Autowired

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy [hunk 7]
```diff
@@ -73,20 +81,53 @@ class ManualJudgmentStage implements StageDefinitionBuilder, AuthenticatedStage
     final long timeout = TimeUnit.DAYS.toMillis(3)
 
     @Autowired(required = false)
-    EchoService echoService
+    private final EchoService echoService
+
+    @Autowired(required = false)
```
Comment (ajordens) [inline orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:86]: This is still not correct.

The `@Autowired` should be on the constructor and not these attributes.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy [hunk 8]
```diff
@@ -104,6 +100,9 @@ class OperationsController {
   @Autowired(required = false)
   Front50Service front50Service
 
+  @Autowired(required = false)
+  private FiatPermissionEvaluator fiatPermissionEvaluator
```
Comment (ajordens) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy:104]: This does not seem needed.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/web/config/WebConfiguration.groovy [hunk 9]
```diff
@@ -54,6 +56,11 @@ class WebConfiguration {
     return  new MetricsInterceptor(registry, "controller.invocations", ["application"], ["BasicErrorController"])
   }
 
+  @Bean
+  ManualJudgmentAuthzGroupsUtil mnlJudgUtil(Optional<Front50Service> front50Service) {
```
Comment (ajordens) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/web/config/WebConfiguration.groovy:60]: Please do not truncate `mnl` and `Judg` when naming beans ... `s/mnlJudgUtl/ManualJugementAuthorizationUtil` (or something.

File: orca-echo/src/main/java/com/netflix/spinnaker/orca/echo/util/ManualJudgmentAuthzGroupsUtil.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+public class ManualJudgmentAuthzGroupsUtil {
+
+  @Autowired(required = false)
+  Front50Service front50Service;
+
+  public ManualJudgmentAuthzGroupsUtil(Optional<Front50Service> front50Service) {
+    this.front50Service = front50Service.orElse(null);
+  }
+
+  public static boolean checkAuthorizedGroups(
```
Comment (ajordens) [inline orca-echo/src/main/java/com/netflix/spinnaker/orca/echo/util/ManualJudgmentAuthzGroupsUtil.java:41]: You still have not added any inline comments here around what this function is intended to do.

I don't quite understand why we're checking against application permissions when the manual judgment could not have happened if the user didn't have EXECUTE.

```
@PreAuthorize("hasPermission(this.getPipeline(#id)?.application, 'APPLICATION', 'EXECUTE')")
@RequestMapping(value = "/pipelines/{id}/stages/{stageId}", method = RequestMethod.PATCH)
PipelineExecution updatePipelineStage(
```

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy [hunk 11]
```diff
@@ -73,20 +81,53 @@ class ManualJudgmentStage implements StageDefinitionBuilder, AuthenticatedStage
...
     @Override
     TaskResult execute(StageExecution stage) {
       StageData stageData = stage.mapTo(StageData)
+      def username = AuthenticatedRequest.getSpinnakerUser().orElse(stage.lastModified ? stage.lastModified.user : "")
+      boolean fiatEnabled = fiatStatus ? fiatStatus.isEnabled() : false
+      boolean isAuthorized = false
+      def appPermissions
+      def stageRoles
+      if (fiatEnabled) {
+        appPermissions = getApplicationPermissions(stage)
```
Comment (ajordens) [inline orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:117]: I'd rather we not introduce a call to `front50` if there aren't any required roles here. The majority of use-cases (and certainly everything pre-existing) will not need this.

**Review by ajordens — CHANGES_REQUESTED** [review]

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy (hunk 8 above)
Comment (sanopsmx) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy:104]: done...

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 7 above)
Comment (sanopsmx) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:86]: I removed, but i find the @Autowired in many places. This should not be a concern.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/web/config/WebConfiguration.groovy (hunk 9 above)
Comment (sanopsmx) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/web/config/WebConfiguration.groovy:60]: done

File: orca-echo/src/main/java/com/netflix/spinnaker/orca/echo/util/ManualJudgmentAuthzGroupsUtil.java (hunk 10 above)
Comment (sanopsmx) [reply orca-echo/src/main/java/com/netflix/spinnaker/orca/echo/util/ManualJudgmentAuthzGroupsUtil.java:41]: Added inline comments.

Application permissions has to be 'CREATE, WRITE, EXECUTE' to proceed further in the manual judgment stage.

This method checks if the logged in user has role in the manual judgment stage authorized groups.
   We fetch the user roles and check if that role is authorized in the manual judgment stage role.
   if the user role exists , then we check with the application permission roles.
   If the application permission role has 'READ' then we return false(not authorized)
   If the application permission role has 'CREATE, EXECUTE, WRITE' then we return true(authorized)

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 11 above)
Comment (sanopsmx) [reply orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy:117]: done..

**Review by ajordens — APPROVED** [review]
Going to merge this down and fast follow-up with a few stylistic tweaks.

I'll tag @sanopsmx on the follow-up PR.

**Comment by ajordens** [comment]
I believe there are still some logical issues with this PR ... going to revert it and let's continue the effort over on https://github.com/spinnaker/orca/pull/4023.

**Comment by sanopsmx** [comment]
@robzienert ,
@ajordens 

When is the spinnaker 1.25 release date? Is this fix going to be in 1.25 release?

**Comment by sanopsmx** [comment]
> @robzienert ,
> @ajordens
> 
> When is the spinnaker 1.25 release date? Is this fix going to be in 1.25 release?

Even this deck fix should be in this release.
https://github.com/spinnaker/deck/pull/8779
